saleor/plugins/avatax/excise/__init__.py

```diff
@@ -0,0 +1,478 @@
+
+
+@dataclass
+class TransactionCreateRequestData:
+    EffectiveDate: str
+    InvoiceDate: str
+    TitleTransferCode: str
+    TransactionType: str
+    TransactionLines: List[TransactionLine]
+    InvoiceNumber: Optional[str] = None
+
+
+def generate_request_data(
+    transaction_type: str,
+    lines: List[TransactionLine],
+    checkout: "Checkout",
+):
+    date = checkout.last_change.strftime("%m/%d/%y")
+    data = TransactionCreateRequestData(
+        EffectiveDate=date,
+        InvoiceDate=date,
+        TitleTransferCode="DEST",
+        TransactionType=transaction_type,
+        TransactionLines=lines,
+    )
+
+    return data
+
+
+def generate_request_data_order(
+    transaction_type: str,
+    lines: List[TransactionLine],
+    order: "Order",
+):
+    date = order.created.strftime("%m/%d/%y")
+    data = TransactionCreateRequestData(
+        EffectiveDate=date,
+        InvoiceDate=date,
+        InvoiceNumber=order.pk,
+        TitleTransferCode="DEST",
+        TransactionType=transaction_type,
+        TransactionLines=lines,
+    )
+
+    return data
+
+
+def get_checkout_lines_data(
+    checkout: "Checkout", discounts=None
+) -> List[TransactionLine]:
+    data: List[TransactionLine] = []
+    lines_info = fetch_checkout_lines(checkout)
+    channel = checkout.channel
+    tax_included = Site.objects.get_current().settings.include_taxes_in_prices
+    shipping_address = checkout.shipping_address
+    if shipping_address is None:
+        raise TaxError("Shipping address required for ATE tax calculation")
+
+    for line_info in lines_info:
+        line = line_info.line
+        variant = line_info.variant
+        product = line_info.product
+        stock = variant.stocks.for_country(shipping_address.country).first()
+        warehouse = stock.warehouse if stock else None
+        cost_price = (
+            line_info.channel_listing.cost_price.amount
+            if line_info.channel_listing.cost_price
+            else None
+        )
+        line_amount = (
+            base_calculations.base_checkout_line_total(
+                line_info,
+                channel,
+                discounts,
+            ).gross.amount
+            if tax_included
+            else base_calculations.base_checkout_line_total(
+                line_info,
+                channel,
+                discounts,
+            ).net.amount
+        )
+
+    data.append(
+        TransactionLine(
+            InvoiceLine=line_info.line.id,
+            ProductCode=variant.sku,
+            UnitPrice=line_info.channel_listing.price.amount,
+            UnitOfMeasure=line_info.product.product_type.get_value_from_private_metadata(
+                get_metadata_key("UnitOfMeasure")
+            ),
+            BilledUnits=Decimal(line.quantity),
+            LineAmount=line_amount,
+            AlternateUnitPrice=cost_price,
+            TaxIncluded=tax_included,
+            UnitQuantity=variant.get_value_from_private_metadata(
```

Comment:
Just to confirm. 
Is it fine if the `UnitQuantity` or `UnitQuantityUnitOfMeasure` will get the `None` value? It is the case when `get_value_from_private_metadata` return None.

Reply:
These fields are undocumented as of right now in ATE :roll_eyes: but testing the API I can confirm they accept null. Both UnitQuantity: null and just leaving out UnitQuantity is fine.

Business requirements may require them and it's possible you could get an error depending on the ATE configuration. But this plugin doesn't need to consider that. If you configure ATE to require these fields for some rule, then you'll get a hopefully readable error.